### PR TITLE
Some support for TOML files

### PR DESCRIPTION
**Pull request by lkishalmi**
This support provides the following:
* lexer based syntax coloring
* useful keybindings
* some code completion based on the keys in the current document. 

image

### CODE REVIEW

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlLexer.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    public TomlLexer(LexerRestartInfo<TomlTokenId> info) {
+        this.tokenFactory = info.tokenFactory();
+        try {
+            this.lexer = new org.tomlj.internal.TomlLexer(new LexerInputCharStream(info.input()));
+            if (info.state() != null) {
+                ((LexerState) info.state()).restore(lexer);
+            }
+        } catch (Throwable ex) {
+            ex.printStackTrace();
```
Comment (mbien) [inline ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlLexer.java:47]: catching Exception would be better, since Throwable includes Error too which is better not to be caught.

handle with `Exceptions.printStackTrace`?

File: ide/languages.toml/nbproject/project.properties [hunk 2]
```diff
@@ -0,0 +1,19 @@
+is.eager=true
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (mbien) [inline ide/languages.toml/nbproject/project.properties:2]: value above license

**Review by mbien — COMMENTED** [review]
awesome! We should use this PR as an up2date example of writing a language support module. 

I quickly looked through it as first pass (no real review) and added two minor comments inline. Good job!

I think it would be good to have at least a basic "black box" test for this which parses a file and checks that nothing explodes.

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlLexer.java (hunk 1 above)
Comment (lkishalmi) [reply ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlLexer.java:47]: Well, I probably should return an ERROR token, than maybe make sure than an EOF is returned. 
The TOML lexre was able to produce an ArrayIndexOutOfBounds exception and IAE on wrong input, though it might be because I was wrong about the state saving. Anyway. Thanks for spotting this one. Need to check what shall I really do.

File: ide/languages.toml/nbproject/project.properties (hunk 2 above)
Comment (lkishalmi) [reply ide/languages.toml/nbproject/project.properties:2]: Yeah, that's bad. I need to patch the UI action that put it there one day.

**Comment by lkishalmi** [comment]
As of tests, I still need to study the Yaml language support how to write them. There could be some edge cases in the keyboard handling and the code completion.

**Comment by Chris2011** [comment]
Great work, will help me also to extend other languages and add more :)

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlCompletionProvider.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    private class TomlCompletionQuery extends AsyncCompletionQuery {
+
+        @Override
+        protected void query(CompletionResultSet resultSet, Document doc, int caretOffset) {
+            Set<String> candidates = new HashSet<>();
+            try {
+                int prefixOfs = keyPrefixOffset(doc, caretOffset);
+                String prefix = doc.getText(prefixOfs, caretOffset - prefixOfs);
```
Comment (sdedic) [inline ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlCompletionProvider.java:64]: Q: is `doc` read-locked at this time ?

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlCompletionProvider.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    private static final Set<TomlTokenId> DOT_OR_KEY = EnumSet.of(TomlTokenId.DOT, TomlTokenId.KEY);
+
+    static int keyPrefixOffset(Document doc, int offset) throws BadLocationException {
+        AbstractDocument d = (AbstractDocument) doc;
+        try {
+            d.readLock();
+            TokenHierarchy th = TokenHierarchy.get(doc);
+            TokenSequence<TomlTokenId> ts = th.tokenSequence();
+            ts.move(offset);
+            ts.movePrevious();
```
Comment (sdedic) [inline ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlCompletionProvider.java:117]: Maybe check the result of `movePrevious`; indicates end(begin)-of-token stream

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlTypedTextInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        ts.move(context.getOffset());
+        ts.movePrevious();
+        while ((ts.token() != null) && STRING_OR_WS.contains(ts.token().id())) {
+            ts.movePrevious();
+        }
+        return (ts.token() != null) && TomlTokenId.ML_STRING_START == ts.token().id();
+    }
+
+    private static int quotesInLine(Context context, char quote) throws BadLocationException {
+        BaseDocument doc = (BaseDocument) context.getDocument();
```
Comment (sdedic) [inline ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlTypedTextInterceptor.java:139]: Use `LineDocument`, no dependency on editor.lib needed.

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/layer.xml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                            <attr name="instanceCreate" methodvalue="org.netbeans.modules.options.editor.spi.CustomizerFactories.createDefaultTabsAndIndentsCustomizerFactory"/>
+                            <attr name="previewTextFile" stringvalue="org/netbeans/modules/languages/toml/TomlExample.toml"/>
+                            <attr name="position" intvalue="100"/>
+                        </file>
+                    </folder>
+                </folder>
+            </folder>
+        </folder>
+    </folder>
+    <!--folder name="Templates">
```
[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.toml/src/org/netbeans/modules/languages/toml/layer.xml]: Resolved with the recent changes.

File: ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlTypedTextInterceptor.java (hunk 5 above)
Comment (lkishalmi) [reply ide/languages.toml/src/org/netbeans/modules/languages/toml/TomlTypedTextInterceptor.java:139]: Good tip thanks!

File: ide/languages.toml/nbproject/project.properties [hunk 7]
```diff
@@ -0,0 +1,19 @@
+is.eager=true
```
[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.toml/nbproject/project.properties]: Actually without that one the commit validation fails, even locally. It seems to be a valid issue (something with ergonomics, if I remember well). Checked the Yaml, yes that module is Regular, so I removed the eager, and try to find another way to fix the commit validation tests. 

Any hints?

[the remark this answers is not in the document]
Comment (mbien) [reply ide/languages.toml/nbproject/project.properties]: is the ergonomics module still loaded during IDE use? I vaguely remember that it had something to do with usage statistics which have been disabled.

[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.toml/nbproject/project.properties]: Yes it does. If you install from zip or Snap packages, the IDE starts with minimal functionality, then the rest is either switched on by ergonomics or directly form the plugins dialog.

[the remark this answers is not in the document]
Comment (mbien) [reply ide/languages.toml/nbproject/project.properties]: oh right! I confused that with something else - my mistake.
